plugins/GSdx/GSOsdManager.cpp

```diff
@@ -0,0 +1,445 @@
+	m_log_speed = m_log_speed < 2 ? 2 : m_log_speed > 10 ? 10 : m_log_speed;
+	m_monitor_enabled = theApp.GetConfigB("osd_monitor_enabled");
+	m_indicator_enabled = theApp.GetConfigB("osd_indicator_enabled");
+	m_osd_transparency = theApp.GetConfigI("osd_transparency");
+	m_osd_transparency = m_osd_transparency < 0 ? 0 : m_osd_transparency > 200 ? 200 : m_osd_transparency;
+
+	if (FT_Init_FreeType(&m_library)) {
+		fprintf(stderr, "Failed to init the freetype library\n");
+		return;
+	}
+
+	LoadFont();
+
+	/* The space character's width is used in GeneratePrimitives() */
+	AddGlyph(' ');
+}
+
+GSOsdManager::~GSOsdManager() {
+	FT_Done_FreeType(m_library);
+}
+
+GSVector2i GSOsdManager::get_texture_font_size() {
+	return GSVector2i(m_atlas_w, m_atlas_h);
+}
+
+void GSOsdManager::upload_texture_atlas(GSTexture* t) {
+	if (!m_face) return;
+
+	if (m_char_info.size() > 96) // we only reserved space for this many glyphs
+		fprintf(stderr, "More than 96 glyphs needed for OSD");
+
+	// This can be sped up a bit by only uploading new glyphs
+	int x = 0;
+	for(auto &pair : m_char_info) {
+		if(FT_Load_Char(m_face, pair.first, FT_LOAD_RENDER)) {
+			fprintf(stderr, "failed to load char U%d\n", (int)pair.first);
+			continue;
+		}
+
+		// Size of char
+		pair.second.ax = m_face->glyph->advance.x >> 6;
+		pair.second.ay = m_face->glyph->advance.y >> 6;
+
+		pair.second.bw = m_face->glyph->bitmap.width;
+		pair.second.bh = m_face->glyph->bitmap.rows;
+
+		pair.second.bl = m_face->glyph->bitmap_left;
+		pair.second.bt = m_face->glyph->bitmap_top;
+
+		GSVector4i r(x, 0, x+pair.second.bw, pair.second.bh);
+		if (r.width())
+			t->Update(r, m_face->glyph->bitmap.buffer, m_face->glyph->bitmap.pitch);
+
+		if (r.width() > m_max_width) m_max_width = r.width();
+
+		pair.second.tx = (float)x / m_atlas_w;
+		pair.second.ty = (float)pair.second.bh / m_atlas_h;
+		pair.second.tw = (float)pair.second.bw / m_atlas_w;
+
+		x += pair.second.bw;
+	}
+
+	m_texture_dirty = false;
+}
+
+#if __GNUC__ < 5 || ( __GNUC__ == 5 && __GNUC_MINOR__ < 4 )
+/* This is dumb in that it doesn't check for malformed UTF8. This function
+ * is not expected to operate on user input, but only on compiled in strings */
+void dumb_utf8_to_utf32(const char *utf8, char32_t *utf32, unsigned size) {
+	while(*utf8 && --size) {
+		if((*utf8 & 0xF1) == 0xF0) {
+			*utf32++ = (utf8[0] & 0x07) << 18 | (utf8[1] & 0x3F) << 12 | (utf8[2] & 0x3F) << 6 | utf8[3] & 0x3F;
+			utf8 += 4;
+		} else if((*utf8 & 0xF0) == 0xE0) {
+			*utf32++ =                          (utf8[0] & 0x0F) << 12 | (utf8[1] & 0x3F) << 6 | utf8[2] & 0x3F;
+			utf8 += 3;
+		} else if((*utf8 & 0xE0) == 0xC0) {
+			*utf32++ =                                                   (utf8[0] & 0x1F) << 6 | utf8[1] & 0x3F;
+			utf8 += 2;
+		} else if((*utf8 & 0x80) == 0x00) {
+			*utf32++ =                                                                           utf8[0] & 0x7F;
+			utf8 += 1;
+		}
+	}
+
+	if(size) *utf32 = *utf8; // Copy NUL char
```

Comment:
When talking about characters it is NUL https://en.wikipedia.org/wiki/Null_character, when talking about pointers it is NULL.